Mathlib/RingTheory/TensorProduct.lean

```diff
@@ -436,28 +436,26 @@ instance : AddMonoidWithOne (A ⊗[R] B) :=
 
 instance : AddCommMonoid (A ⊗[R] B) := by infer_instance
 
+instance instMul : Mul (A ⊗[R] B) := ⟨fun a b ↦ mul a b⟩
+
+instance : NonUnitalNonAssocSemiring (A ⊗[R] B) := {
+  -- porting note : `left_distrib` and `right_distrib` are proved by `simp` in mathlib3
+  -- See https://github.com/leanprover-community/mathlib4/issues/5026
+  -- Probably because `mul` is defined to be bi-R-linear and then coerced to function?
+  left_distrib := fun a b c => show mul a (b + c) = mul a b + mul a c by simp
+  right_distrib := fun a b c => show mul (a + b) c = mul a c + mul b c by simp
+  zero_mul := fun a => show mul 0 a = 0 by simp
+  mul_zero := fun a => show mul a 0 = 0 by simp
+}
+
+instance instNonUnitalSemiring : NonUnitalSemiring (A ⊗[R] B) := {
+  mul_assoc := mul_assoc
+}
+
 instance instSemiring : Semiring (A ⊗[R] B) :=
-  { (by infer_instance : AddMonoidWithOne (A ⊗[R] B)),
```

Comment:
I worry that this discarded the `natCast` field, though I have no idea whether it actually does.

Reply:
Do you still worry given that it compiles and the benchmark results?

Reply:
Yes, because neither of those steps necessarily checks for diamonds problems, and previously we relied on the source code making it obvious that they didn't exist. I'll add an `example` to verify that it's fine.

Reply:
Thanks. I don't understand the issue that you're concerned about so please feel free to check this out (or explain the issue to me, either will do).

Reply:
The issue is
```lean
example : (algebraNat : Algebra ℕ (ℕ ⊗[ℕ] ℕ)) = leftAlgebra := rfl -- fails
```
but this already failed on master so you're off the hook!